framework/src/media/demux/mpeg2ts/ParseManager.cpp

```diff
@@ -0,0 +1,278 @@
+	auto pPMTInstance = pPMTParser->getPMTInstance(progNum);
+
+	if (pPMTInstance && pPMTInstance->isCompleted()) {
+		size_t i;
+		size_t num = pPMTInstance->numOfElementary();
+		for (i = 0; i < num; i++) {
+			auto pStream = pPMTInstance->getPMTElementary(i);
+			if (!pStream) {
+				meddbg("Run out of memory!\n");
+				return false;
+			}
+
+			switch (pStream->getStreamType()) {
+				case PMTElementary::STREAM_TYPE_AUDIO_AAC:   // fall through
+				case PMTElementary::STREAM_TYPE_AUDIO_MPEG2: // fall through
+				case PMTElementary::STREAM_TYPE_AUDIO_AC3:   // fall through
+				case PMTElementary::STREAM_TYPE_AUDIO_MPEG1: // fall through
+				case PMTElementary::STREAM_TYPE_AUDIO_HE_AAC:
+					streamType = pStream->getStreamType();
+					pid = pStream->getElementaryPID();
+					medvdbg("stream type 0x%02x, pid 0x%x\n", streamType, pid);
+					return true;
+				default:
+					break;
+			}
+		}
+		meddbg("Can not find audio stream in PMT of program number %d\n", progNum);
+		return false;
+	}
+
+	meddbg("PMT of program number %d has not be received!\n", progNum);
+	return false;
+}
+
+bool ParserManager::getPrograms(std::vector<prog_num_t> &programs)
+{
+	size_t i, num;
+
+	auto pPATParser = std::static_pointer_cast<PATParser>(getParser(PATParser::TABLE_ID));
+	if (!pPATParser) {
+		meddbg("PAT parser is not found!\n");
+		return false;
+	}
+
+	if (!pPATParser->isRecv()) {
+		meddbg("PAT has not been received yet!\n");
+		return false;
+	}
+
+	num = pPATParser->sizeOfProgram();
+	for (i = 0; i < num; i++) {
+		programs.push_back(pPATParser->getProgramNumber(i));
+	}
+
+	return true;
+}
+
+bool ParserManager::syncProgramInfoFromPAT(void)
+{
+	size_t i, num;
+	ts_pid_t pid;
+	prog_num_t progNum;
+	std::map<int, ts_pid_t> pmt_elements;
+
+	auto pPATParser = std::static_pointer_cast<PATParser>(getParser(PATParser::TABLE_ID));
+	if (!pPATParser) {
+		meddbg("PAT parser is not found!\n");
+		return false;
+	}
+
+	if (!pPATParser->isRecv()) {
+		meddbg("PAT has not been received yet!\n");
+		return false;
+	}
+
+	// Clear current PMT Pids
+	mPMTPids.clear();
+
+	// Update new informations from PAT
+	num = pPATParser->sizeOfProgram();
+	for (i = 0; i < num; ++i) {
+		progNum = pPATParser->getProgramNumber(i);
+		pid = pPATParser->getProgramMapPID(progNum);
+		if ((pid != (ts_pid_t)INVALID_PID) && (progNum != (prog_num_t)PATParser::NETWORK_PID_PN)) {
+			int key = PMTParser::makeKey(pid, progNum);
+			pmt_elements[key] = pid;
+			mPMTPids.push_back(pid);
+			medvdbg("index %d, pmt pid 0x%02x\n", i, pid);
+		}
+	}
+
+	auto pPMTParser = std::static_pointer_cast<PMTParser>(getParser(PMTParser::TABLE_ID));
+	if (!pPMTParser) {
+		meddbg("PMT parser is not found!\n");
+		return false;
+	}
+	// Reinitialize PMT parser and update new PMT elements to PMT parser
+	pPMTParser->Initialize();
+	pPMTParser->updatePMTElements(pmt_elements);
+	return true;
+}
+
+bool ParserManager::isPMTPid(ts_pid_t pid)
+{
+	auto iter = mPMTPids.begin();;
```

Comment:
1. There are two ;
2. How about using ranged-for?
```c++
for (auto iter : mPMPids) {
    if (*iter == pid) {
        return true;
    }
}
``

Reply:
fixed, thanks